amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnector.scala

```diff
@@ -73,6 +71,8 @@ private[amqp] trait AmqpConnectorLogic { this: GraphStageLogic =>
 
   /** remember to call if overriding! */
   override def postStop(): Unit = {
+    connection.removeShutdownListener(shutdownListener)
+    channel.removeShutdownListener(shutdownListener)
```

Comment:
Heh, apparently I put up a PR a few minutes after this one, but in either case: the channel shouldn't require special shutdown listener removal as it is always going to be distinct to this instance. My understanding is that it's the connection that persists, but `channel.close()` should always *actually* close and therefore unlink it's registered shutdown listeners. (Anecdotally in my heap analysis when opening this issue, channel shutdown listeners were not an issue.)

Reply:
Yeah. Channels are always unique to a stage.

My understanding is that the suhtdownListener is needed just in case the channel is closed by the server or for whatever other reason, so the stage gets informed and fails itself.

Reply:
Agreed that the shutdown listener is necessary for channels. I'm suggesting that the explicit removal of the channel shutdown listener may be unnecessary.

Reply:
Oh yeah! you are right on that one.
It's not really necessary. Just being extra-careful on that one!

Reply:
I'm not entirely convinced 'extra careful' is a good thing here. Explicitly removing a listener could have some runtime impact, especially if the listener would be naturally removed on channel shutdown anyway and if the listener collection is a thread-safe or locking collection (which, presumably, it is).

I think in either case, regardless of approach taken, I'm nervous about the absence of test coverage here. Perhaps we should brainstorm a way to test this? I think it could be done with a proxying connection that verifies only a single channel shutdown listener is added despite re-using a cached connection?

Reply:
What side runtime effect do you expect?

Removing a channel shutdown is literally:
```
    @Override
    public void removeShutdownListener(ShutdownListener listener)
    {
        synchronized(this.monitor) {
            this.shutdownListeners.remove(listener);
        }
    }
```
so thread-safe and no exceptions thrown.

Having this explicitly it's more future proof imho since we don't know how this connector will evolve in the future.

I like the changes that you have done to #887. However, it's now an incomplete subset of this fix since it doesn't tackle the duplicated listeners. So I'd still go with this fix.

But I agree 100% on the issue around testing disconnect. That's a known issue in this connector and the MQTT one.
Using a TCP proxy was done on the MQTT connector (see https://github.com/akka/alpakka/blob/master/mqtt/src/test/scala/akka/stream/alpakka/mqtt/scaladsl/MqttSourceSpec.scala#L256) but AFAIK it's not a 100% reliable test.

I'm in EU so not much overlapping TZ with you, but happy to help with the testing if I can :)

Reply:
> What side runtime effect do you expect?

`synchronized` will result in locking and slower memory access. So I'm advocating that we avoid unnecessary calls to mutate this collection, especially behind a lock.

> However, it's now an incomplete subset of this fix since it doesn't tackle the duplicated listeners

I feel like we're just not communicating on this, or perhaps just not agreeing on the problem space. There is no current issue with duplicated channel listeners, only connection listeners. The potential for error with shared channels is also very low and represents a misuse of ampq channels, or at least a misuse of alpakka amqp graph stages at the least. 

Perhaps it's more of a personal ethos, but I generally dislike making changes to code that isn't exhibiting a known problem and has no supported use case that would exhibit it. 

This PR represents more change to as-of-yet untested code. The reason I've left my PR up so far is I think it will both solve the immediate problem and the narrow footprint of the change should allow it to be more easily tested with confidence in all changes.

I will ping again when I have a testing proposal in place. As of now, I feel like the two of us are basically at an impasse as to which is the best approach to take to solving #883